gnd/src/main/java/com/google/android/gnd/persistence/remote/firestore/schema/AuditInfoConverter.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.android.gnd.persistence.remote.firestore.schema;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+import com.google.android.gnd.model.AuditInfo;
+import com.google.android.gnd.model.Mutation;
+import com.google.android.gnd.model.User;
+import com.google.firebase.Timestamp;
+import java.util.Date;
+import java8.util.Optional;
+
+/** Converts between Firestore nested objects and {@link AuditInfo} instances. */
+class AuditInfoConverter {
+
+  /**
+   * Converts a POJO representing user and timestamp data in Firebase into an equivalent model
+   * object. This should never be empty in Firebase, but this method returns a default value when
+   * the input is null to support legacy or corrupt dbs.
+   */
+  @NonNull
+  static AuditInfo toAuditInfo(@Nullable AuditInfoNestedObject doc) {
+    if (doc == null || doc.getClientTimeMillis() == null) {
```

Comment:
If we know the user, shouldn't we ascribe the audit to the user even if the time is somehow unknown?

Reply:
We could also return defaults from the AuditInfoNestedObject getters themselves -- since we're accounting for a peculiarity of firebase and not our model. 

Plus that ensures the work performed by the converter is actual conversion from store values to model values (timestamp->date) and we don't give it the double duty of defining defaults for store values.

Reply:
I don't have a particularly strong opinion on this though, so up to you.

Reply:
> We could also return defaults from the AuditInfoNestedObject getters themselves -- since we're accounting for a peculiarity of firebase and not our model.

This would split out part of the conversion of remove data, making the conversion harder to debug.

> Plus that ensures the work performed by the converter is actual conversion from store values to model values (timestamp->date) and we don't give it the double duty of defining defaults for store values.

I think it makes sense to a canonical representation of what's actually in Firebase separate from the handling of that data. It makes things easier to debug, and consolidates all logic that transforms that data (defaults, fallbacks, or otherwise) into a single place.

> If we know the user, shouldn't we ascribe the audit to the user even if the time is somehow unknown?

Actually now that I think about it, an entity without a time should probably just be considered invalid, since it's one of the only identifying attributes of features and observations. Users, on the other hand, could delete their accounts, meaning we would need to delete the user data.

Reply:
Sounds good!

Reply:
Fixed! Relevant patch: ab1b06fb643e4de3e4d029cedbc93f575535674d